### PR TITLE
Shape inference of simple operations in LazyTensor.

### PR DESCRIPTION
**Pull request by bgogul**
This PR incorporates shape tracking in lazy tensor by using the shape inference functions associated with the TF operations.

### CODE REVIEW

File: Sources/TensorFlow/Core/Tensor.swift [hunk 1]
```diff
@@ -51,10 +51,15 @@ extension Tensor: AnyTensor {
 
 public extension Tensor {
     /// The number of dimensions of the `Tensor`.
-    @inlinable
+    // @inlinable
```
Comment (eaplatanios) [inline Sources/TensorFlow/Core/Tensor.swift]: Why remove the `@inlinable`?

Comment (bgogul) [reply Sources/TensorFlow/Core/Tensor.swift]: Just a temporary hack as `LazyTensorHandle` is not public. We need to make it `@usableFromInline`.

**Comment by bgogul** [comment]
Folks, this PR is ready for review.

File: Sources/TensorFlow/Core/LazyTensorOperation.swift [hunk 2]
```diff
@@ -67,12 +67,29 @@ class LazyTensorHandle: _AnyTensorHandle {
     }
 
     /// The number of dimensions of the underlying `Tensor`.
-    @inlinable
-    var rank: Int { _tfeTensorHandle.rank }
+    @usableFromInline
+    var rank: Int {
+        @_semantics("autodiff.nonvarying")
+        get { return shape.rank }
```
[the remark this answers is not in the document]
Comment (bgogul) [reply Sources/TensorFlow/Core/LazyTensorOperation.swift]: Unfortunately, that is not possible with the TF API. In any case, in both the eager and lazy paths, we will have the shapes available anyway?

**Comment by bgogul** [comment]
Gentle ping. Can someone take a look at this PR, please?

File: Sources/TensorFlow/Core/LazyTensorContext.swift [hunk 3]
```diff
@@ -62,12 +62,22 @@ class LazyTensorOperationsTracker {
 
 struct LazyTensorContext {
     private var operationsTracker = LazyTensorOperationsTracker()
+    private var _shapeTrackingEnabled = true
```
Comment (rxwei) [inline Sources/TensorFlow/Core/LazyTensorContext.swift]:
```suggestion
    private var isShapeTrackingEnabled = true
```

No need for underscore in a less-than-public type.

File: Sources/TensorFlow/Core/LazyTensorContext.swift [hunk 4]
```diff
@@ -62,12 +62,22 @@ class LazyTensorOperationsTracker {
...
 
     static private var threadLocalContext: LazyTensorContext {
```
Comment (rxwei) [inline Sources/TensorFlow/Core/LazyTensorContext.swift]: Consider renaming to just `local`. It's a convention used in a lot of other Swift APIs.

File: Sources/TensorFlow/Core/LazyTensorContext.swift [hunk 5]
```diff
@@ -62,12 +62,22 @@ class LazyTensorOperationsTracker {
...
-        _ThreadLocalState.local.lazyTensorContext
+        _read { yield _ThreadLocalState.local.lazyTensorContext }
+        _modify { yield &_ThreadLocalState.local.lazyTensorContext }
     }
 
     static var operationsTracker: LazyTensorOperationsTracker {
         return threadLocalContext.operationsTracker
     }
+
+    /// A flag that determines whether we should track shapes.
```
Comment (rxwei) [inline Sources/TensorFlow/Core/LazyTensorContext.swift]: Fit to 100 columns? This seems to be using an 80-column limit.

File: Sources/TensorFlow/Core/LazyTensorContext.swift [hunk 6]
```diff
@@ -62,12 +62,22 @@ class LazyTensorOperationsTracker {
...
+    /// We will need to disable shape tracking within certain contexts.
+    /// e.g., we won't be able to compute shapes when tracing.
+    static var shapeTrackingEnabled: Bool {
```
Comment (rxwei) [inline Sources/TensorFlow/Core/LazyTensorContext.swift]:
```suggestion
    static var isShapeTrackingEnabled: Bool {
```

**Pull request merged**